### PR TITLE
Speed Up Variational Strategy

### PR DESCRIPTION
**Pull request by kayween**
This PR includes two improved variational strategy implementations. Both are optimized for large batch stochastic training.

1. A custom autograd function is implemented for the existing `VariationalStrategy`, which saves a few large matrix operations in the backward pass. This achieves up to 3.08x speed up. 
1. A separate class `LargeBatchVariationalStrategy` that is optimized for large-scale training on data center GPUs (e.g., A100). This achieves up to 4.45x speed up.
1. Both implementations by default return the predictive variance (instead of the full covariance) in train mode. We added a new argument `diag` to control this variance vs. covariance behavior. Downstream variational strategies could set `diag=False` if they require the full covariance in train mode.

## Benckmark Results
https://wandb.ai/kayween/gpytorch-fast-variational/reports/PR-Benchmark-for-Variational-Strategy---VmlldzoxNTQzNjY4NQ?accessToken=hr169mp64hcf3u9uhhng9bho0fqbog1ew7s5jx3nay4hde0szq9ated264pyhuby

The test MAE and test NLL are generally the same up to the third digit after the decimal point, which suggests that the two new implementations are numerically stable.

The running time bar plot shows that they achieve 3.08x and 4.45x speed-ups respectively when training SVGPs on the house electric dataset.

### CODE REVIEW

**Comment by kayween** [comment]
Attaching the benchmark results here https://wandb.ai/kayween/gpytorch-fast-variational/reports/PR-Benchmark-for-Variational-Strategy---VmlldzoxNTQzNjY4NQ?accessToken=hr169mp64hcf3u9uhhng9bho0fqbog1ew7s5jx3nay4hde0szq9ated264pyhuby

The test MAE and test NLL are generally the same up to the third digit after the decimal point, which suggests that the two new implementations are numerically stable.

The running time bar plot shows that they achieve 3.08x and 4.45x speed-ups respectively when training SVGPs on the house electric dataset.

**Comment by kayween** [comment]
It seems that `LMCVariationalStrategy` and `OrthogonallyDecoupledVariationalStrategy` still require the full covariance in train mode. However, `VariationalStrategy` only computes the predictive variance in train mode (not the full covariance matrix), which fails some of the test cases.

**Comment by jacobrgardner** [comment]
Let's keep only computing the diagonal in train mode but add a predictive covariance method that eval mode uses and that the other strategies can inherit I guess?

**Comment by kayween** [comment]
> Let's keep only computing the diagonal in train mode but add a predictive covariance method that eval mode uses and that the other strategies can inherit I guess?

I am not entirely sure if this would resolve the test failures? For example, `LMCVariationalStrategy` does not inherit `VariationalStrategy`. It actually takes a variational strategy in its constructor.

https://github.com/cornellius-gp/gpytorch/blob/60be9533fb571482a4ea270921cb635d71da56f8/gpytorch/variational/lmc_variational_strategy.py#L108-L115

**Comment by kayween** [comment]
I think we should always return the full predictive covariance (even in train mode). I can modify the custom autograd function to accomodate that.

Instead of only computing the diagonal, the custom autograd function could return two tensors
```
left = K_XZ K_ZZ^{-1/2}
right = (S - I) @  K_ZZ^{-1/2} K_XZ
```
and then the variational strategy constructs the predictive covariance `K_XX + MatmulLinearOperator(left, right)`. We do need to tweak the custom backward implementation a bit to accomodate this change.

**Comment by jacobrgardner** [comment]
This feels like it'll potentially make it harder for the LargeBatch version to inherit from this though, since it explicitly avoids computing left? Also makes the custom function a bit weirder.

What seems cleaner to me is to pull the else branch that computed the full predictive covar into a separate helper than the train mode one, and then standard SVGP eval and these downstream strategies can use that.

So, predictive_updates and full_predicfive_updates or something. There's no reason ODSVGP can't call self.model.full_predictive_updates instead of self.model directly.

**Comment by kayween** [comment]
> This feels like it'll potentially make it harder for the LargeBatch version to inherit from this though, since it explicitly avoids computing left? Also makes the custom function a bit weirder.

Agreed that it makes the custom autograd function weirder. What particularly bugs me is that this approach would destroy the symmetry in the computation. (As a result, I think PyTorch would allocate a `m x n` tensor in the backward pass that we don't need.) But, if we really want to, we can still make it work for the large batch version. We just need to redefine the left and right tensors.

> So, predictive_updates and full_predicfive_updates or something. There's no reason ODSVGP can't call self.model.full_predictive_updates instead of self.model directly.

I see. This should solve the issue.

A related idea to this. Instead of defining new methods, how do we feel about using a context manager to control this full covariance behavior?
```
with settings.variational_strategy_compute_full_covariance.on():
    model(x)  # return the full predictive covariance regardless of train/eval mode

with settings.variational_strategy_compute_full_covariance.off():  
    model(x)  # compute variance in train mode, and compute full covariance in eval mode
```

It seems that this would change fewer lines of code. A bigger benefit is that the context manager would make fixing `UnwhitenedVariationalStrategy` much easier, which also only computes the predictive variance in train mode.

**Comment by jacobrgardner** [comment]
We could do that. The one issue with settings is that the settings are generally user facing, and I can't imagine a user ever really wanting to change these (e.g., train mode for normal SVGPs should always operate in diag mode, and eval mode should operate in full mode). I think my preference there would just be helper methods.

**Comment by jacobrgardner** [comment]
The torchscript example failures make sense I think -- just need to update the saved model @kayween

File: gpytorch/variational/independent_multitask_variational_strategy.py [hunk 1]
```diff
@@ -49,11 +51,19 @@ def variational_params_initialized(self):
...
+        self,
+        x: Tensor,
+        task_indices: Optional[LongTensor] = None,
+        prior: bool = False,
+        diag: bool = True,
+        **kwargs,
+    ):
         r"""
         See :class:`LMCVariationalStrategy`.
         """
```
[the remark this answers is not in the document]
Comment (jacobrgardner) [reply gpytorch/variational/independent_multitask_variational_strategy.py:64]: @kayween you both define an explicit diag kwarg and check whether it's in kwargs? I assume you meant to just leave it out of the method signature? diag won't be in kwargs if you make it an explicit keyword argument.

[the remark this answers is not in the document]
Comment (kayween) [reply gpytorch/variational/independent_multitask_variational_strategy.py:64]: @jacobrgardner I didn't do the check. This check is copilot's suggestion, which I don't think should be committed.

**Comment by kayween** [comment]
> This now looks good to me, pending (1) diag kwarg oddness in subclasses that don't use it (see my and copilot's comment), (2) the example failure / re-save the trace model, (3) some docstring typos.

(1): I kept the `diag` argument explicit in all classes so that these methods share a consistent signature with parent classes. Removing it in subclasses that don't use it is more intuitive. I am fine either way as long as removing them doesn't break the test cases.

@jacobrgardner what's your thoughts on `UnwhitenedVariationalStrategy` and `NNVariationalStrategy`? These two classes return the predictive varaince instead of the full covariance. So we need to modify them to use the diag arument. I added the `diag` argument but didn't modify their `foward` methods. Because I think these modifications are better addressed in a separate PR. (In particular, `NNVariationalStrategy` would require some efforts because it returns diagonal covariances even in eval mode.) 

(2) and (3): Will do.

File: gpytorch/variational/independent_multitask_variational_strategy.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jacobrgardner) [reply gpytorch/variational/independent_multitask_variational_strategy.py:64]: Oh I see, yep that confused me.

Actually, now that I see it, the copilot suggestion (which also removes diag as a kwarg) is reasonable in the sense that the diag kwarg does just get silently ignored otherwise.

**Comment by jacobrgardner** [comment]
I think leave Unwhitened and NN as is for now. Imo, clean up diag in the subclasses assuming the tests pass and the example failure and then this is good to go.

File: gpytorch/variational/batch_decoupled_variational_strategy.py [hunk 2]
```diff
@@ -181,6 +181,7 @@ def forward(
         inducing_points: Tensor,
         inducing_values: Tensor,
         variational_inducing_covar: Optional[LinearOperator] = None,
+        diag: bool = True,
```
[the remark this answers is not in the document]
Comment (kayween) [reply gpytorch/variational/batch_decoupled_variational_strategy.py:184]: @jacobrgardner We cannot remove this `diag` even though it's unused.

This method will subsequently call model's foward, where `self.model` is an approximate GP. If we don't keep `diag` explicit, it would be passed to the model via `kwargs`. This would eventually trigger an unexpected argument error.
https://github.com/cornellius-gp/gpytorch/blob/3dd3745533c84a6653dd8137cd3c7b483d2d1280/gpytorch/variational/batch_decoupled_variational_strategy.py#L194

We have to remove `diag` from `kwargs` because `self.model.foward` is an user-defined function and we have zero control on it. Thus, we have to either keep `diag` explicitly, or pop it from kwargs as copilot suggested.

Among the two choices, I actually prefer keeping `diag` explicitly because it's simpler

File: gpytorch/variational/variational_strategy.py [hunk 3]
```diff
@@ -177,12 +260,69 @@ def pseudo_points(self) -> Tuple[Tensor, Tensor]:
...
+        r"""Compute the predictive mean and covariance updates. Adding the return values of this method to the prior
+        mean and covariance yields the predictive mean and covariance.
+
+        The predictive mean update is `K_{XZ} K_{ZZ}^{-1/2} m`.
+
+        The predictive covariance update is `K_{XZ} K_{ZZ}^{-1/2} (S - I) K_{ZZ}^{-1/2} K_{ZX}`.
+
+        :param chol: The Cholesky factor `K_{ZZ}^{-1/2}`.
+        :param induc_data_covar: The covariance between the inducing points and the data `K_{ZX}`.
+        :param middle_term: The difference between the variational and prior covariances `S - I`.
```
[the remark this answers is not in the document]
Comment (kayween) [reply gpytorch/variational/variational_strategy.py]: fixed

File: gpytorch/variational/variational_strategy.py [hunk 4]
```diff
@@ -51,6 +51,89 @@ def _ensure_updated_strategy_flag_set(
...
+        d_chol = -2.0 * inv_chol_times_middle @ d_middle
+
+        # Then add the derivative of `K_ZZ^{-1/2}` received from the predictive mean
+        d_chol = d_chol - inv_chol_times_inducing_values @ interp_term_times_dmean.mT
+
+        # NOTE: In principle, we need to zero out the lower triangular part because `chol` is lower triangular. It is
+        # actually not necessary here, because `d_chol` is immediately fed into `cholesky_backward`, which does not
+        # care about the upper triangular part. We keep it here for consistency with PyTorch's implementation.
+        # https://github.com/pytorch/pytorch/blob/4a0693682a8574bdc36e1ca2ea7bd2ddf5c19340/torch/csrc/autograd/FunctionsManual.cpp#L1999-L2003
+        # NOTE: If we want to get fansy, fusing this backward with `cholesky_backward` will save a matmul. It may not
```
[the remark this answers is not in the document]
Comment (kayween) [reply gpytorch/variational/variational_strategy.py]: typo fixed

File: gpytorch/variational/variational_strategy.py [hunk 5]
```diff
@@ -177,12 +260,69 @@ def pseudo_points(self) -> Tuple[Tensor, Tensor]:
...
+        mean and covariance yields the predictive mean and covariance.
+
+        The predictive mean update is `K_{XZ} K_{ZZ}^{-1/2} m`.
+
+        The predictive covariance update is `K_{XZ} K_{ZZ}^{-1/2} (S - I) K_{ZZ}^{-1/2} K_{ZX}`.
+
+        :param chol: The Cholesky factor `K_{ZZ}^{-1/2}`.
+        :param induc_data_covar: The covariance between the inducing points and the data `K_{ZX}`.
+        :param middle_term: The difference between the variational and prior covariances `S - I`.
+        :param inducing_values: The whitened variational mean `m`.
```
[the remark this answers is not in the document]
Comment (kayween) [reply gpytorch/variational/variational_strategy.py]: added docstring describing `diag`

[the remark this answers is not in the document]
Comment (kayween) [reply gpytorch/variational/variational_strategy.py]: fixed

**Comment by kayween** [comment]
@jacobrgardner All test cases are passed now. The trace mode error was fixed by making a linear operator dense.

Now the only thing remaining is what do we do about those unused `diag`. Not all unused `diag` can be removed because this argument could be accidentially passed down to `model.forward` via `kwargs`, which could trigger unexpected argument errors. For example, this happend in `BatchDecoupledVariationalStrategy`.

To properly handle this, we could manually pop `diag` from `kwargs`. The fundamental problem here is that most variational strategies don't use this argument (with a few exceptions like `_VariationalStrategy`, `VariationalStrategy`, the unwhitened version, and the nearest neighbor). To be on the safe side, we need to do a lot of manual pop, which doesn't seem to a pretty solution.

If we feel strong about those unused `diag`, we should probably switch back to the context manager approach?

**Comment by jacobrgardner** [comment]
No, I prefer unused diag in some cases to a context manager. I think this is ready to go!
